### PR TITLE
fix(core): align STOMP importance ratios with Gumbel behavior policy

### LINKED ISSUE #2136: fix(core): align STOMP importance ratios with Gumbel behavior policy
## Summary

STOMP's intra-option behavior selector and its importance-ratio calculation disagree for near-tied Q-values.

`_select_action_epsilon_greedy` selects the non-exploratory action with:

```python
argmax(q_values + 1e-6 * Gumbel(0, 1))
```

Its greedy-component probabilities are therefore `softmax(q_values / 1e-6)`. However, `_clipped_epsilon_greedy_importance_ratio` calls `_epsilon_greedy_action_probabilities`, which treats every value within absolute tolerance `1e-6` of the maximum as an exactly uniform tie.

For `q_values = [0, 5e-7]`, behavior epsilon `0.2`, and target epsilon `0`:

- actual Gumbel greedy probabilities: approximately `[0.37754068, 0.62245935]`
- correct target/behavior ratios: approximately `[0.9390799, 1.0409585]`
- current helper ratios: `[1.0, 1.0]`

The wrong ratio is consumed by `_update_intra_option_policy`, scaling eligibility traces, weight updates, and the differential average-reward update with probabilities from a different policy than generated the action.

## Proposed fix

Compute the greedy-component probabilities from the exact Gumbel-max distribution, then mix in uniform epsilon exploration. Add a regression test for near-tied values and retain exact-tie uniform behavior.

This is a measurement/control correctness fix; it does not claim benchmark improvement or scientific promotion.

### PR DESCRIPTION
**Pull request by agent-cortex**
## Summary

Measurement/control correctness fix in the STOMP option learner
(`alberta_framework/core/options.py`). **No benchmark, performance, or
promotion claim is made** — this only repairs an importance ratio that was
scaling learning updates with probabilities from a policy that never generated
the action.

Closes #2136.

## What was broken and its measurement consequence

STOMP's intra-option / base behavior selectors break ties with the Gumbel-max
trick:

```python
greedy = jnp.argmax(q_vals + 1e-6 * jr.gumbel(noise_key, (n_actions,)))
```

By the Gumbel-max trick the greedy-component probability of action *i* is
`softmax(q_vals / 1e-6)_i` — **not** a uniform split over near-tied Q-values.

But `_epsilon_greedy_action_probabilities` modeled the greedy component as a
uniform split over every value within `atol=1e-6` of the max:

```python
greedy_mask = jnp.isclose(q, max_q, atol=1e-6, rtol=0.0)   # -> uniform split
```

`_clipped_epsilon_greedy_importance_ratio` calls that helper for both the
behavior and target policy, so `_update_intra_option_policy` scaled eligibility
traces, weight updates, and the differential average-reward update with an
importance ratio computed from a policy that did **not** generate the action.

Worked example — `q_values = [0.0, 5e-7]`, `behavior_epsilon = 0.2`,
`target_epsilon = 0.0`:

| quantity | correct (Gumbel-max, temp `1e-6`) | old buggy helper |
|---|---|---|
| target greedy probs | `softmax([0, 0.5]) ≈ [0.37754068, 0.62245935]` | `[0.5, 0.5]` |
| behavior probs | `≈ [0.40203, 0.59797]` | `[0.5, 0.5]` |
| ratios `target/behavior` | `≈ [0.9390799, 1.0409585]` | `[1.0, 1.0]` |

The old helper silently returned unit ratios for near-ties, corrupting the
off-policy correction exactly where Q-values are close.

## The fix

- Introduce a module-level `_GUMBEL_TIEBREAK_TEMPERATURE = 1.0e-6` and use it in
  **all three** selectors (`_select_action_epsilon_greedy`,
  `_select_action_epsilon_greedy_from_q`,
  `_select_action_epsilon_greedy_from_q_masked`) and in the probability helper,
  so the sampled behavior policy and its importance-ratio model can never drift
  apart again. Replacing the inline `1e-6` literals with the constant is a pure
  no-op (same float value); selector draws are bit-identical (verified).
- Compute the greedy component as the exact numerically-stable softmax
  `jax.nn.softmax((q - max_q) / temp)`. Subtracting `max_q` before dividing
  keeps large Q-gaps saturating to a near one-hot argmax instead of overflowing;
  exact ties still reduce to the uniform split. dtype stays `float32`; the public
  signature and `_clipped_epsilon_greedy_importance_ratio` API are unchanged.

Single module, single lane, failing-test-first.

## Failing-then-passing reproduction

New test `tests/test_stomp_importance_ratio.py` (5 tests, ≥10 lines of real
assertions): near-tie softmax split, near-tie importance ratio, exact-tie
uniform retention, well-separated one-hot collapse with selector parity, and
the shared-constant invariant.

### (a) FAILING against the unmodified uniform helper

Command: `.venv/bin/python -m pytest tests/test_stomp_importance_ratio.py -q -o addopts=""`
(helper temporarily reverted to the old `jnp.isclose` uniform logic):

```
FF...                                                                    [100%]
=================================== FAILURES ===================================
______________ test_near_tie_probabilities_follow_gumbel_softmax _______________
E       Mismatch at indices:
E        [0]: 0.5 (ACTUAL), 0.37754068 (DESIRED)
E        [1]: 0.5 (ACTUAL), 0.62245935 (DESIRED)
tests/test_stomp_importance_ratio.py:32: AssertionError
____________ test_near_tie_importance_ratio_matches_behavior_policy ____________
E       Mismatch at indices:
E        [0]: 1.0 (ACTUAL), 0.9390799 (DESIRED)
E        [1]: 1.0 (ACTUAL), 1.0409585 (DESIRED)
tests/test_stomp_importance_ratio.py:53: AssertionError
=========================== short test summary info ============================
FAILED tests/test_stomp_importance_ratio.py::test_near_tie_probabilities_follow_gumbel_softmax
FAILED tests/test_stomp_importance_ratio.py::test_near_tie_importance_ratio_matches_behavior_policy
2 failed, 3 passed in 3.62s
exit=1
```

### (b) PASSING after the fix

Command: `.venv/bin/python -m pytest tests/test_stomp_importance_ratio.py -v -o addopts=""`:

```
tests/test_stomp_importance_ratio.py::test_near_tie_probabilities_follow_gumbel_softmax PASSED [ 20%]
tests/test_stomp_importance_ratio.py::test_near_tie_importance_ratio_matches_behavior_policy PASSED [ 40%]
tests/test_stomp_importance_ratio.py::test_exact_ties_still_split_uniformly PASSED [ 60%]
tests/test_stomp_importance_ratio.py::test_well_separated_q_collapses_to_argmax PASSED [ 80%]
tests/test_stomp_importance_ratio.py::test_temperature_constant_is_shared_with_selectors[1e-06] PASSED [100%]
5 passed in 4.00s
exit=0
```

## Verification commands and results

- Touched-file tests — `.venv/bin/python -m pytest tests/test_stomp_importance_ratio.py tests/test_stomp_reward_semantics.py tests/test_options.py -q -o addopts=""`:
  ```
  113 passed in 31.18s
  exit=0
  ```
- Lint — `.venv/bin/python -m ruff check .`:
  ```
  All checks passed!
  ```
- Types — `.venv/bin/python -m mypy` (strict, py312):
  ```
  Success: no issues found in 224 source files
  ```

No pinned `outputs/` artifact was touched; `options.py` is not a registered
evidence source, so `alberta-evidence-status` outcomes are unaffected.

## What remains open

This fix only aligns the importance-ratio model with the already-existing Gumbel
behavior policy. It does not claim any downstream benchmark, learning, or
promotion effect; the reference-life scorecard and IPMNIST campaigns are
untouched.

> Evidence note: the immutable `user-attachments` URL minting tool
> (`attach.mjs`) could not authenticate its browser session in this environment
> (GitHub returned "Incorrect username or password" for the stored browser
> credential; `gh` uses a separate token). The failing-then-passing and
> verification logs are therefore pasted inline above verbatim, bound to the
> head SHA in the marker below.

- [x] logs: inline in this PR body (see "Failing-then-passing reproduction" and "Verification commands and results" above)

AI provider/model: anthropic / claude-opus-4-8
Client / agent tooling: pi-coding-agent
Contribution skill revision: elizaOS/slopdotcash@51ee9f1b1e88ced844e28ae09184c65e8bed223c:skills/contribute-to-asi
Attribution status: self-reported
— [eliza-swarm]

### CODE REVIEW

**Review by deepanshu-yd — APPROVED** [review]
Reproduced: `pytest tests/test_stomp_importance_ratio.py -v` 5 passed with fix `6f7c5c68`.

Diff `alberta_framework/core/options.py:56-68,997-1070` introduces `_GUMBEL_TIEBREAK_TEMPERATURE=1e-6` and replaces inline `1e-6` in 3 selectors + probability helper. Helper now `softmax((q-max_q)/temp)` vs old `isclose(q,max_q,atol=1e-6)` uniform split — matches Gumbel-max `softmax(q/temp)` exact. Verified near-tie `[0,5e-7]` gives `softmax([0,0.5])=[0.3775,0.6224]` not `[0.5,0.5]`, ratio `[0.939,1.040]` not `[1,1]`; exact ties stay uniform, well-separated collapses to one-hot with selector parity 16 seeds.

Without fix test module fails import of new constant (expected); PR body failing logs show 2/5 fail with old uniform helper.

`ruff check` pass, `mypy` success 224. No-evidence lane, single module, failing-test-first. Approving — closes #2136.

**Review by hermesagent270-commits — APPROVED** [review]
Reviewed at `6f7c5c68eb9525bed44a60783070859c35ea9f0c` (static review only).

The change aligns the modeled greedy-component probabilities with the selectors' actual Gumbel-max distribution: `argmax(q + temperature * Gumbel)` induces `softmax(q / temperature)`. Reusing one constant across all three selectors and the ratio helper prevents future scale drift, and centering by the maximum before division preserves the distribution while avoiding overflow for large finite values. Exact ties remain uniform, the epsilon mixture is unchanged, and the tests cover the near-tie ratio regression, sign/mass behavior, exact ties, and the hard-greedy regime. This is a measurement/control correctness fix with appropriately nonpromoting scope.

No static blocker found. Only triage has run on GitHub at this head, so the repository's required lint, typing, package, portability, and runtime checks must run and pass before merge.

**Comment by lalalune** [comment]
The intended probability correction is retained in #2808, which also centers the actual Gumbel selectors (including masked selection). This version leaves actual sampling uncentered, so finite common Q offsets can swallow the 1e-6 noise in float32 while the modeled probabilities still claim a stochastic tie. Closing the incomplete variant in favor of the tested centered implementation; three regressions fail before that fix and the 206-test CBP/options panel passes.
